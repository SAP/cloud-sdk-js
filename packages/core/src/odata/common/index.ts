export * from './batch-response';
export * from './edm-types';
export * from './entity';
export * from './entity-builder';
export * from './filter';
export * from './odate-batch-consts';
export * from './order';
export * from './request';
export * from './request-builder';
export * from './selectable';
export * from './time';
export * from './uri-conversion';
