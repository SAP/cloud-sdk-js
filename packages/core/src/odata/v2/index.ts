/* Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. */

export * from './edm-types';
export * from './entity';
export * from './entity-deserializer';
export * from './entity-serializer';
export * from './filter-functions';
export * from './odata-batch-change-set';
export * from './odata-batch-request-builder';
export * from './odata-batch-retrieve-request';
export * from './payload-value-converter';
export * from './request-builder';
export * from './selectable/custom-field';
export * from './uri-conversion';
export * from './legacy-request-configs';
export * from './extract-odata-etag';
export * from '../common';
