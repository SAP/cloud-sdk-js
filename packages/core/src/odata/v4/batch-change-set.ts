/* Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. */
import { v4 as uuid } from 'uuid';
import { BatchChangeSet } from '../common/request-builder/batch/batch-change-set';
import { EntityV4 } from './entity';
import {
  CreateRequestBuilderV4,
  DeleteRequestBuilderV4,
  UpdateRequestBuilderV4
} from './request-builder';

/**
 * @deprecated Since v1.30.0. Use [[BatchChangeSet]] directly
 * Representation of a batch change set, which holds a collection of write operations.
 */
export class ODataBatchChangeSetV4<
  RequestT extends
    | CreateRequestBuilderV4<EntityV4>
    | UpdateRequestBuilderV4<EntityV4>
    | DeleteRequestBuilderV4<EntityV4>
> implements BatchChangeSet<RequestT> {
  /**
   * @deprecated Since v1.30.0. Use [[boundary]] instead.
   */
  get changeSetId() {
    return this.boundary;
  }

  /**
   * Create an instance of ODataBatchChangeSetV4.
   * @param requests Requests to combine to one change set.
   * @param boundary Boundary used in the multipart request.
   */
  constructor(
    readonly requests: RequestT[],
    readonly boundary: string = `changeset_${uuid()}`
  ) {}
}
