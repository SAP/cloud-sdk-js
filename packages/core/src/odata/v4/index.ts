/* Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. */

export * from './edm-types';
export * from './entity';
export * from './entity-deserializer';
export * from './entity-serializer';
export * from './odata-batch-change-set';
export * from './odata-batch-request-builder';
export * from './odata-batch-retrieve-request';
export * from './payload-value-converter';
export * from './request-builder';
export * from './selectable/custom-field';
export * from './uri-conversion';
export * from './extract-odata-etag';
export * from '../common';
export * from '../common/selectable/one-to-many-link';
