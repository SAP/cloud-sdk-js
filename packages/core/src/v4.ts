/* Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. */

/**
 * @packageDocumentation
 * @experimental The v4 module is related to OData v4 and is still experimental. Use with caution.
 */
export * from './scp-cf';
export * from './util';
export * from './odata/v4';
export * from './header-builder';
export * from './http-agent';
export * from './http-client';
