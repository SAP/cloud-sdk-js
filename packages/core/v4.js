/* Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. */
'use strict';
function __export(m) {
  for (const p in m) {
    if (!exports.hasOwnProperty(p)) {
      exports[p] = m[p];
    }
  }
}
Object.defineProperty(exports, '__esModule', { value: true });
/**
 * @packageDocumentation
 * @experimental The v4 module is related to OData v4 and is still experimental. Use with caution.
 */
__export(require('./dist/v4'));
// # sourceMappingURL=v4.js.map
