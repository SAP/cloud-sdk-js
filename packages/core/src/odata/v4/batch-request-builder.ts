/* Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. */

import { errorWithCause } from '@sap-cloud-sdk/util';
import { v4 as uuid } from 'uuid';
import {
  Destination,
  DestinationNameAndJwt,
  DestinationOptions,
  toDestinationNameUrl
} from '../../scp-cf';
import { Constructable, BatchResponse } from '../common';
import { MethodRequestBuilderBase } from '../common/request-builder/request-builder-base';
import { ODataBatchRequestConfig } from '../common/request/odata-batch-request-config';
import { EntityV4 } from './entity';
import {
  ODataBatchChangeSetV4,
  serializeBatchRequest
} from './batch-request-serializer';
import {
  CreateRequestBuilderV4,
  DeleteRequestBuilderV4,
  GetAllRequestBuilderV4,
  GetByKeyRequestBuilderV4,
  UpdateRequestBuilderV4
} from './request-builder';
import { parseBatchResponse } from './batch-response-parser';

/**
 * Create a batch request to invoke multiple requests as a batch. The batch request builder accepts retrieve requests, i. e. [[GetAllRequestBuilder | getAll]] and [[GetByKeyRequestBuilder | getByKey]] requests and change sets, which in turn can contain [[CreateRequestBuilder | create]], [[UpdateRequestBuilder | update]] or [[DeleteRequestBuilder | delete]] requests.
 * The retrieve and change sets will be excuted in order, while the order within a change set can vary.
 */
export class ODataBatchRequestBuilderV4 extends MethodRequestBuilderBase<
  ODataBatchRequestConfig
> {
  /**
   * Creates an instance of ODataBatchRequestBuilder.
   *
   * @param defaultServicePath - Service path
   * @param requests - An array of retrieve requests or change sets
   * @param entityToConstructorMap - A map that holds the entity type to constructor mapping
   */
  constructor(
    readonly defaultServicePath: string,
    readonly requests: (
      | ODataBatchChangeSetV4<
          | CreateRequestBuilderV4<EntityV4>
          | UpdateRequestBuilderV4<EntityV4>
          | DeleteRequestBuilderV4<EntityV4>
        >
      | GetAllRequestBuilderV4<EntityV4>
      | GetByKeyRequestBuilderV4<EntityV4>
    )[],
    readonly entityToConstructorMap: Record<string, Constructable<EntityV4>>
  ) {
    super(new ODataBatchRequestConfig(defaultServicePath, uuid()));
    this.requestConfig.payload = serializeBatchRequest(this);
  }

  /**
   * Execute the given request and return the according promise. Please notice: The sub-requests may fail even the main request is successful.
   *
   * @param destination - Targeted destination on which the request is performed.
   * @param options - Options to employ when fetching destinations.
   * @returns Promise resolving to the requested data.
   */
  async execute(
    destination: Destination | DestinationNameAndJwt,
    options?: DestinationOptions
  ): Promise<BatchResponse[]> {
    return this.build(destination, options)
      .then(request => request.execute())
      .then(response =>
        parseBatchResponse(response, this.entityToConstructorMap)
      )
      .catch(error =>
        Promise.reject(
          errorWithCause(
            `The batch request failed! The destination provided: ${toDestinationNameUrl(
              destination
            )} and the options: ${options}.`,
            error
          )
        )
      );
  }
}
