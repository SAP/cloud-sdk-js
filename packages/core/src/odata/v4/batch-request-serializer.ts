/* Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. */

import {
  serializeChangeSet,
  serializeRequest
} from '../common/request-builder/batch/batch-request-serializer';
import { BatchChangeSet } from '../common/request-builder/batch/batch-change-set';
import { EntityV4 } from './entity';
import {
  CreateRequestBuilderV4,
  DeleteRequestBuilderV4,
  GetAllRequestBuilderV4,
  GetByKeyRequestBuilderV4,
  UpdateRequestBuilderV4
} from './request-builder';

/**
 * @deprecated Since v1.30.0. Use [[serializeRequest]] instead.
 * Build a string as the request body of the retrieve request.
 * Below is an example of the generated body, where the two empty line are mandatory to make the request valid.
 * *** example starts ***
 * Content-Type: application/http
 * Content-Transfer-Encoding: binary
 *
 * GET /SomeUrl/API_BUSINESS_PARTNER/A_BusinessPartnerBank?$format=json&$top=1 HTTP/1.1
 *
 *
 * *** example ends ***
 * @param request - The request builder of the retrieve request.
 * @returns The request body.
 */
export function toBatchRetrieveBodyV4(
  request: GetAllRequestBuilderV4<EntityV4> | GetByKeyRequestBuilderV4<EntityV4>
): string {
  return serializeRequest(request);
}

/**
 * @deprecated Since v1.30.0. Use [[serializeChangeSet]] instead.
 * Serialize change set to string.
 * @param changeSet - Change set containing a collection of write operations.
 * @returns The serialized string representation of a change set.
 */
export function toBatchChangeSetV4<
  T extends
    | CreateRequestBuilderV4<EntityV4>
    | UpdateRequestBuilderV4<EntityV4>
    | DeleteRequestBuilderV4<EntityV4>
>(changeSet: BatchChangeSet<T>): string | undefined {
  return serializeChangeSet(changeSet);
}
