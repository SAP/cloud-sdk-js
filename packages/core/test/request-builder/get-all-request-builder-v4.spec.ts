/* Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. */
import { GetAllRequestBuilder } from '../../src/odata/v4';
import { muteLoggers } from '../test-util/mute-logger';
import {
  defaultDestination,
  mockGetRequest,
  unmockDestinationsEnv
} from '../test-util/request-mocker';
import { TestEntity } from '../test-util/test-services/v4/test-service';
import {
  createOriginalTestEntityData1,
  createOriginalTestEntityData2,
  createOriginalTestEntityDataWithLinks,
  createTestEntityV4
} from '../test-util/test-data';

describe('GetAllRequestBuilder', () => {
  let requestBuilder: GetAllRequestBuilder<TestEntity>;

  beforeAll(() => {
    muteLoggers('http-agent', 'destination-accessor', 'environment-accessor');
  });

  afterEach(() => {
    unmockDestinationsEnv();
  });

  beforeEach(() => {
    requestBuilder = new GetAllRequestBuilder(TestEntity);
  });

  describe('url', () => {
    it('is built correctly', async () => {
      const expected =
        '/testination/sap/opu/odata/sap/API_TEST_SRV/A_TestEntity?$format=json';
      const actual = await requestBuilder.url(defaultDestination);
      expect(actual).toBe(expected);
    });
  });

  describe('execute', () => {
    it('returns all entities', async () => {
      const testEntity1 = createOriginalTestEntityData1();
      const testEntity2 = createOriginalTestEntityData2();

      mockGetRequest(
        {
          responseBody: { value: [testEntity1, testEntity2] }
        },
        TestEntity
      );

      const actual = await new GetAllRequestBuilder(TestEntity).execute(
        defaultDestination
      );
      expect(actual).toEqual([
        createTestEntityV4(testEntity1),
        createTestEntityV4(testEntity2)
      ]);
    });

    it('top(1) returns the first entity', async () => {
      const testEntity1 = createOriginalTestEntityData1();
      mockGetRequest(
        {
          query: { $top: 1 },
          responseBody: { value: [testEntity1] }
        },
        TestEntity
      );

      const actual = await requestBuilder.top(1).execute(defaultDestination);
      expect(actual).toEqual([createTestEntityV4(testEntity1)]);
    });

    it('skip(1) skips the first entity', async () => {
      const testEntity2 = createOriginalTestEntityData2();
      mockGetRequest(
        {
          query: { $skip: 1 },
          responseBody: { value: [testEntity2] }
        },
        TestEntity
      );
      const actual = await requestBuilder.skip(1).execute(defaultDestination);
      expect(actual).toEqual([createTestEntityV4(testEntity2)]);
    });

    it('should resolve when ALL_FIELDS is selected and links are expanded', async () => {
      const testEntity = createOriginalTestEntityDataWithLinks();
      mockGetRequest(
        {
          query: {
            $select: '*',
            $expand: 'to_SingleLink,to_MultiLink'
          },
          responseBody: { value: [testEntity] }
        },
        TestEntity
      );
      const actual = await requestBuilder
        .select(TestEntity.ALL_FIELDS)
        .expand(TestEntity.TO_SINGLE_LINK, TestEntity.TO_MULTI_LINK)
        .execute(defaultDestination);
      expect(actual).toEqual([createTestEntityV4(testEntity)]);
    });
  });
});
