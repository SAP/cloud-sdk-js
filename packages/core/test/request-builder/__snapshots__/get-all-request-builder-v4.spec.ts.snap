// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GetAllRequestBuilder execute throws an error when request execution fails 1`] = `"get request to /testination/missing failed! "`;

exports[`GetAllRequestBuilder execute throws an error when the destination cannot be found 1`] = `"Could not find a destination with name \\"NonExistentDestination\\"! Unable to execute request."`;
